src/deploy/functions/ensure.ts

```diff
@@ -0,0 +1,131 @@
+  return new FirebaseError(`Cloud Functions deployment requires the Cloud Build API to be enabled. The current credentials do not have permission to enable APIs for project ${clc.bold(
+    projectId
+  )}.
+
+Please ask a project owner to visit the following URL to enable Cloud Build:
+
+https://console.cloud.google.com/apis/library/cloudbuild.googleapis.com?project=${projectId}
+
+For additional information about this requirement, see Firebase FAQs:
+${FAQ_URL}
+`);
+}
+
+function isPermissionError(e: { context?: { body?: { error?: { status?: string } } } }): boolean {
+  return e.context?.body?.error?.status === "PERMISSION_DENIED";
+}
+
+/**
+ * Checks for various warnings and API enablements needed based on the runtime
+ * of the deployed functions.
+ *
+ * @param projectId Project ID upon which to check enablement.
+ */
+export async function ensureCloudBuildEnabled(projectId: string): Promise<void> {
+  try {
+    await ensure(projectId, CLOUD_BUILD_API, "functions");
+  } catch (e: any) {
+    if (isBillingError(e)) {
+      throw nodeBillingError(projectId);
+    } else if (isPermissionError(e)) {
+      throw nodePermissionError(projectId);
+    }
+
+    throw e;
+  }
+}
+
+// We previously force-enabled AR. We want to wait on this to see if we can give
+// an upgrade warning in the future. If it already is enabled though we want to
+// remember this and still use the cleaner if necessary.
+export async function maybeEnableAR(projectId: string): Promise<boolean> {
+  if (previews.artifactregistry) {
+    return ensureApiEnabled.check(
+      projectId,
+      "artifactregistry.googleapis.com",
+      "functions",
+      /* silent= */ true
+    );
+  }
+  await ensureApiEnabled.ensure(projectId, "artifactregistry.googleapis.com", "functions");
+  return true;
+}
+
+/**
+ * Ensures that runtime service account has access to the secrets.
+ *
+ * To avoid making more than one simultaneous call to setIamPolicy calls per secret, the function batches all
+ * service account that requires access to it.
+ */
+export async function ensureSecretAccess(b: backend.Backend) {
+  const ensureAccess = async (projectId: string, secret: string, serviceAccounts: string[]) => {
+    logLabeledBullet(
+      "functions",
+      `ensuring ${clc.bold(serviceAccounts.join(", "))} access to ${clc.bold(secret)}.`
+    );
+    await ensureServiceAgentRole(
+      { name: secret, projectId },
+      serviceAccounts,
+      "roles/secretmanager.secretAccessor"
+    );
+    logLabeledSuccess(
+      "functions",
+      `ensured ${clc.bold(serviceAccounts.join(", "))} access to ${clc.bold(secret)}.`
+    );
+  };
+
+  // Collect all service accounts that requires access to a secret.
+  // projectId -> secretName -> Set of service accounts
+  const toEnsure: Record<string, Record<string, Set<string>>> = {};
+  for (const e of backend.allEndpoints(b)) {
+    const sa = e.serviceAccountEmail || defaultServiceAccount(e.project);
```

Comment:
Again, we're going to need to use the right service account for e.platform

Reply:
Also, I think we have this backwards in a dangerous way. It's very possible that I'd run a privileged function as a separate service account, but this logic grants all known service accounts to all known environment variables.

Reply:
hmm I don't think I understand what you are trying to say here.

This code:

1. Constructs a mapping from secret to all service account that needs access to the secret
2. Maybe SetIamPolicy to allow each SA associated w/ a secret to access the secret

IIUC, you are worried about possibly sending IAM policy update to allow all SAs identified in the backend to access all secrets referenced in the source. It doesn't do that.

Reply:
this code snippet probably needs a new review - it's gone through several changes:

1) default service account logic works for both v1 and v2.
2) It looks at existing function definitions to avoid making IAM calls if secret is already bound to a SA.